### PR TITLE
OpenMM radians; factors of 2

### PR DESCRIPTION
**Pull request by leeping**
Hi Jason,

This pull request fixes a few things in agreement between Gromacs and OpenMM (for just one water molecule).  The problems appeared to be:

1) Gromacs parser for "angles" and "bonds" were off by a factor of two in the force constant
2) Degrees were being passed to OpenMM's angle force, but it expected radians

More to come soon.
- Lee-Ping

### CODE REVIEW

**Comment by leeping** [comment]
Before:

```
#========================================================#
#|              GROMACS energy components               |#
#========================================================#
Bond                      [ 1.638048] 
Angle                     [ 0.014803] 
LJ-(SR)                   [ 0.] 
Coulomb-(SR)              [ 0.] 
Potential                 [ 1.652851] 
----------------------------------------------------------
#========================================================#
#|         OpenMM energy components via ParmEd          |#
#========================================================#
HarmonicBondForce         3.27609537205 
HarmonicAngleForce        6622479.3397 
NonbondedForce            0.0 
CMMotionRemover           0.0 
Potential                 6622482.61579 
----------------------------------------------------------
Energy Difference (kJ/mol):
6622480.96294
RMS Force Difference (kJ/mol/nm):
1637524.78646
```

After:

```
Defines: {'FLEXIBLE': 1}
Constraints None
Running without a target, using current directory.
The 'gmxpath' option was not specified; using default.
#========================================================#
#|              GROMACS energy components               |#
#========================================================#
Bond                      [ 1.638048] 
Angle                     [ 0.014803] 
LJ-(SR)                   [ 0.] 
Coulomb-(SR)              [ 0.] 
Potential                 [ 1.652851] 
----------------------------------------------------------
#========================================================#
#|         OpenMM energy components via ParmEd          |#
#========================================================#
HarmonicBondForce         1.63804768603 
HarmonicAngleForce        0.0148031110171 
NonbondedForce            0.0 
CMMotionRemover           0.0 
Potential                 1.65285079704 
----------------------------------------------------------
Energy Difference (kJ/mol):
-2.02956623951e-07
RMS Force Difference (kJ/mol/nm):
2.08266710064e-12
```

**Comment by swails** [comment]
Oh man! I knew about that factor of 2, too. Can you check the in-line parameters? I think I got the factor of 2 there, but I might have missed the angle conversion. 

Also, I think that AngleType takes the angle in radians, so the conversion inside omm_angle_force will break CHARMM and Amber energies/forces. It may make sense to change that, but for now let's leave AngleType alone and convert to radians in the gromacs parser. 

But this is great! Thanks!

**Comment by swails** [comment]
Or at least I got the factor of 2 for the bond type. Maybe not the angle and Urey-Bradley.

**Comment by leeping** [comment]
Hi Jason,

I tried converting to radians in the Gromacs parser (and leaving the createSystem()) alone because the latter option can break the functionality with other programs.  However, it seemed that `AngleType` always returns the angle in degrees, so degrees gets passed to `createSystem()` regardless of what I do.

I'm currently looking at nonbonded forces for a cluster of TIP4P water molecules.  It looks like the electrostatic energy matches but there's something up with Lennard-Jones.

Thanks,
- Lee-Ping

**Comment by leeping** [comment]
Hi Jason,

The latest changes include a modification to the "copy" method in Atom to copy over the rmin and epsilon parameters if they're present.  Previously it was just copying the charge.

Also a factor of 2 was missing in the sigma -> rmin conversion.

Before:

```
#========================================================#
#|              GROMACS energy components               |#
#========================================================#
LJ-(SR)                   [ 23.715706] 
Coulomb-(SR)              [-131.993503] 
Potential                 [-108.277797] 
----------------------------------------------------------
#========================================================#
#|         OpenMM energy components via ParmEd          |#
#========================================================#
HarmonicBondForce         6.07725465317e-12 
NonbondedForce            -131.993509576 
CMMotionRemover           0.0 
Potential                 -131.993509576 
----------------------------------------------------------
Energy Difference (kJ/mol):
-23.7157125763
RMS Force Difference (kJ/mol/nm):
236.094109572
```

After:

```
#========================================================#
#|              GROMACS energy components               |#
#========================================================#
LJ-(SR)                   [ 23.715706]
Coulomb-(SR)              [-131.993503]
Potential                 [-108.277797]
----------------------------------------------------------
#========================================================#
#|         OpenMM energy components via ParmEd          |#
#========================================================#
HarmonicBondForce         6.07725465317e-12
NonbondedForce            -108.277803795
CMMotionRemover           0.0
Potential                 -108.277803795
----------------------------------------------------------
Energy Difference (kJ/mol):
-6.79535027359e-06
RMS Force Difference (kJ/mol/nm):
0.00139966728419
```

Thanks,
- Lee-Ping

**Comment by leeping** [comment]
Also, I think I added the appropriate factors of two for the in-line parameters.  A few of them were already there.

**Comment by leeping** [comment]
Hmm, it turns out I didn't have to modify any methods in your low-level topology objects.  A couple of lines in `parametrize()` did the trick:

```
@@ -679,6 +679,8 @@ class GromacsTopologyFile(Structure):
             types.claim()
         # Assign all of the parameters. If they've already been assigned (i.e.,
         # on the parameter line itself) keep the existing parameters
+        for atom in self.atoms:
+            atom.atom_type = params.atom_types[atom.type]
         for bond in self.bonds:
```

**Comment by leeping** [comment]
Energies and forces now match for a small peptide in the gas phase (includes proper dihedrals).  The fix was very simple, I had to set `scee` and `scnb` when setting dihedrals from dihedraltypes.

I got an error in printing out a dihedral type - I wasn't sure whether self.phase should be a number or Quantity, so I just changed the `__repr__` function.

```
@@ -2155,7 +2155,7 @@ class DihedralType(_ListItem, _ParameterType):

     def __repr__(self):
         return ('<%s; phi_k=%.3f, per=%d, phase=%.3f, scee=%.3f, scnb=%.3f>' %
-                (type(self).__name__, self.phi_k, self.per, self.phase,
+                (type(self).__name__, self.phi_k, self.per, _strip_units(self.phase, u.degree),
                  self.scee, self.scnb))
```

Before:

```
#========================================================#
#|              GROMACS energy components               |#
#========================================================#
Bond                      [ 1.530821] 
Angle                     [ 6.580444] 
Proper-Dih.               [ 80.37973] 
LJ-14                     [ 45.572594] 
Coulomb-14                [ 388.997589] 
LJ-(SR)                   [-19.834761] 
Coulomb-(SR)              [-689.862488] 
Potential                 [-186.636078] 
----------------------------------------------------------
#========================================================#
#|         OpenMM energy components via ParmEd          |#
#========================================================#
HarmonicBondForce         1.53079992448 
HarmonicAngleForce        6.58044884425 
PeriodicTorsionForce      80.3797142083 
NonbondedForce            -151.754802568 
CMMotionRemover           0.0 
Potential                 -63.2638395914 
----------------------------------------------------------
Energy Difference (kJ/mol):
123.372238409
RMS Force Difference (kJ/mol/nm):
166.872703001
```

After:

```
#========================================================#
#|              GROMACS energy components               |#
#========================================================#
Bond                      [ 1.5308] 
Angle                     [ 6.580449] 
Proper-Dih.               [ 80.379714] 
LJ-14                     [ 45.572583] 
Coulomb-14                [ 388.997686] 
LJ-(SR)                   [-19.834736] 
Coulomb-(SR)              [-689.862447] 
Potential                 [-186.635951] 
----------------------------------------------------------
#========================================================#
#|         OpenMM energy components via ParmEd          |#
#========================================================#
HarmonicBondForce         1.53079992448 
HarmonicAngleForce        6.58044884425 
PeriodicTorsionForce      80.3797142083 
NonbondedForce            -275.126928766 
CMMotionRemover           0.0 
Potential                 -186.635965789 
----------------------------------------------------------
Energy Difference (kJ/mol):
-1.4789320943e-05
RMS Force Difference (kJ/mol/nm):
2.40511926185e-05
```

**Comment by swails** [comment]
OK, all of the docstrings say "degrees" and the unit conversions are to degrees.  If you _don't_ supply units, then it doesn't attempt any conversion, and the Amber and CHARMM parsers both went to radians.  So it was a disagreement between the docs and the code (ugh).  Thanks for this.  I'll go ahead and merge this PR and then update the places where I used radians.

Hopefully this doesn't screw anybody up.

File: chemistry/gromacs/gromacstop.py [hunk 1]
```diff
@@ -593,7 +593,7 @@ def read(self, fname, defines=None, parametrize=True):
                         phase = float(words[5]) * u.degrees
                         phi_k = float(words[6]) * u.kilojoules_per_mole
                         per = int(words[7])
-                        dt = DihedralType(phi_k, per, phase)
+                        dt = DihedralType(phi_k, per, phase, scee=1/self.defaults.fudgeQQ, scnb=1/self.defaults.fudgeLJ)
```
Comment (swails) [inline chemistry/gromacs/gromacstop.py:596]: Good catch!

File: chemistry/topologyobjects.py [hunk 2]
```diff
@@ -2155,7 +2155,7 @@ def __eq__(self, other):
 
     def __repr__(self):
         return ('<%s; phi_k=%.3f, per=%d, phase=%.3f, scee=%.3f, scnb=%.3f>' %
-                (type(self).__name__, self.phi_k, self.per, self.phase,
+                (type(self).__name__, self.phi_k, self.per, _strip_units(self.phase, u.degree),
```
Comment (swails) [inline chemistry/topologyobjects.py:2158]: This should've been done in the constructor (my bad), but I'll fix it post-merge.

**Comment by leeping** [comment]
No worries.

I am now testing DHFR in the gas phase, and it says some dihedral parameters are missing.  I'm looking into the wildcards.
- Lee-Ping

**Pull request merged**